libraries/chain/block_log.cpp

```diff
@@ -615,6 +628,68 @@ namespace eosio { namespace chain {
             FC_LOG_AND_RETHROW()
          }
 
+         std::vector<char> read_serialized_block_by_num(uint32_t block_num, verify_block_num_t verify_block_num) final {
+            try {
+               uint64_t pos = get_block_pos(block_num);
+
+               if (pos == block_log::npos || !head) {
+                  // Rare case. No need a special code for it.
+                  // Just fall back to the regular `retry_read_block_by_num` and then serialize.
+                  auto ret = retry_read_block_by_num(block_num);
+                  if (ret) {
+                     return fc::raw::pack(*ret);
+                  } else {
+                     return {};
+                  }
+               }
+
+               if (verify_block_num == verify_block_num_t::yes) {
+                  auto bnum = get_block_num_at(block_file, pos);
+                  if (block_num != bnum) {
+                     wlog("Wrong block num was read from block log. expected block_num: ${n}, returned block num: ${bnum}",
+                           ("n", block_num)("bnum", bnum));
+                     return {};
+                  }
+               }
+
+               uint64_t block_size     = 0;
+               uint32_t head_block_num = block_header::num_from_id(head->id);
+               constexpr uint32_t block_pos_size = sizeof(uint64_t); // size of block position field in the block log file
+
+               if (block_num < head_block_num) {
+                  // current block is not the last block in the log file.
+                  uint64_t next_block_pos = get_block_pos(block_num + 1);
+                  if (next_block_pos == block_log::npos) {
+                     wlog("no position found for block ${n}", ("n", block_num + 1));
+                     return {};
+                  }
+                  if (next_block_pos <= pos) {
+                     wlog("next block position ${nn} should be greater than current block position ${cn}", ("nn", next_block_pos)("cn", pos));
+                     return {};
+                  }
```

Comment:
I think this can be `assert(next_block_pos > pos);`

Reply:
It was decided to use EOS_ASSERT for preventing undefined behavior on our side and for consistency.